### PR TITLE
Split edpm into nodeset  and deployment stages

### PR DESCRIPTION
**Pull request by fao89**

### CODE REVIEW

File: examples/va/hci/edpm-pre-ceph/kustomization.yaml
Comment (abays) [inline examples/va/hci/edpm-pre-ceph/kustomization.yaml:1]: Can we get rid of this file?  The automation will directly be using the `nodeset` and `deployment` sub-dirs, so I think it would be confusing to have this as another option (especially if we're using the sub-dirs to avoid a problem caused by applying `OpenStackDataPlaneNodeSet` and `OpenStackDataPlaneDeployment` at the same time).

Comment (fao89) [reply examples/va/hci/edpm-pre-ceph/kustomization.yaml:1]: yes, let me remove it

File: automation/vars/default.yaml
Comment (abays) [inline automation/vars/default.yaml:1]: We'll have to modify this file to use the split data plane approach

File: automation/vars/uni01alpha.yaml
Comment (abays) [inline automation/vars/uni01alpha.yaml:1]: I think we'll also have to modify this file to use split data plane

**Comment by abays** [comment]
We'll also need to update the various example data plane `.md` files to reflect the new approach

File: automation/vars/uni01alpha.yaml
Comment (fao89) [reply automation/vars/uni01alpha.yaml:1]: Apparently, it is fine, but how can I test this?

Comment (abays) [reply automation/vars/uni01alpha.yaml:1]: I guess I'm confused, because I thought we wanted to break the dataplane stages into two stages, such that `OpenStackDataPlaneNodeSet` can reach `Ready` state in the first stage and then `OpenStackDataPlaneDeployment` reaches `Ready` in the second stage (which we check for via the separate stages' `wait_conditions`, i.e. `oc wait osdpns openstack-edpm -n openstack --for condition=Ready --timeout=1200s` and then `oc wait osdpd edpm-deployment -n openstack --for condition=Ready --timeout=1200s`).  If we don't separate these into two different stages, won't we be applying `OpenStackDataPlaneNodeSet` and `OpenStackDataPlaneDeployment` at the same time still?

File: automation/vars/default.yaml
Comment (cjeanner) [inline automation/vars/default.yaml:34]: hmm if we keep it like that, it's still one stage. I expected to get another block to ensure the nodeset is on its own, something like:
```YAML
- path: examples/va/hci/edpm-pre-ceph/nodeset
  wait_conditions:
    - oc -n openstack wait <nodeset is ready/up-to-date/whatever>
  values:
    - name: edpm-nodeset-values
      src_file: values.yaml
  build_output: nodeset.yaml
- path: examples/va/hci/edpm-pre-ceph/deployment
  wait_conditions:
    - oc -n openstack wait <dataplane is ready/up-to-date/whatever>
  values:
    - name: edpm-deployment-values
      src_file: values.yaml
  build_output: dataplane.yaml
```
or something like that... ?

File: automation/vars/uni01alpha.yaml
Comment (fao89) [reply automation/vars/uni01alpha.yaml:1]: yes, you are right!

File: examples/va/hci/dataplane-post-ceph.md
Comment (abays) [inline examples/va/hci/dataplane-post-ceph.md:40]: `s/nodeset/deployment`

File: va/hci/edpm-post-ceph/kustomization.yaml
Comment (abays) [inline va/hci/edpm-post-ceph/kustomization.yaml:1]: I think it would be better to rewrite using the `Component` pattern like `va/hci/edpm-pre-ceph/kustomization.yaml`, so that we're not duplicating `replacement` logic here and in `va/hci/edpm-post-ceph/deployment/kustomization.yaml` and `va/hci/edpm-post-ceph/nodeset/kustomization.yaml`.  I know there's also a `components` include here of `- ../../../lib/control-plane`, but perhaps it would be okay to place it in `va/hci/edpm-post-ceph/nodeset/kustomization.yaml`?  Or maybe we need a separate stage all together for something like `va/hci/control-plane-post-ceph` to house the control plane update (and thus the EDPM stage would really become three stages: `control-plane`, `nodeset` and `deployment`)?

Comment (fao89) [reply va/hci/edpm-post-ceph/kustomization.yaml:1]: the duplicated `replacements` was a leftover from my previous attempt

**Comment by cjeanner** [comment]
From my side, I'm ready. I want to get some reviews from the team, but https://github.com/openstack-k8s-operators/ci-framework/pull/1547 works.

Since we'll break current jobs downstream if they inject any hook or kustomization (usually they stage_2, which would now be stage_3), I stacked a second patch in the PR to allow using a "better" name, built on the `build_output`.

**Review by fultonj — COMMENTED** [review]
+2

Overall I'm fine with this change.

Before we merge, though someone (because we don't yet have fully automated gating) should ensure that a full ci-framework reproducer run with the updated `automation/vars/default.yaml` completes successfully before we merge. Testing va1 and uni01alpha seems prudent. Let's make sure someone from ci-team agrees before merging.

With VA1 we usually get:
```
├── dataplane-post-ceph.yaml
├── dataplane-pre-ceph.yaml
```
After this PR merges we'll get:
```
├── dataplane-deployment-pre-ceph.yaml
├── dataplane-nodeset-pre-ceph.yaml
├── deployment-post-ceph.yaml
└── nodeset-post-ceph.yaml
```
I built both and then compared the new pre ceph from this branch against the old:
```
cd new
(cat dataplane-deployment-pre-ceph.yaml ; echo "---"; cat dataplane-nodeset-pre-ceph.yaml) > pre.yaml
cd ..
diff -u old/dataplane-pre-ceph.yaml new/pre.yaml 
```
It looked fine (i.e. no meaningful difference).

I compared the new post ceph from this branch against the main branch (old) too. `diff -u old/dataplane-post-ceph.yaml new/nodeset-post-ceph.yaml` is the same as deployment-post-ceph.yaml; which is how it should be.

Nit: I didn't like the name `dataplane-post-ceph.yaml` in VA1 since it also included an update to the control plane. Similarly, nodeset-post-ceph.yaml has a nodeset but it also has an update to the control plane CR, so I think calling it `post-ceph.yaml` would be better. We don't have to block this on that however.

**Comment by psathyan** [comment]
uni01alpha testing results

```
PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
compute-0                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
compute-1                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
localhost                  : ok=388  changed=120  unreachable=0    failed=0    skipped=141  rescued=1    ignored=0   

Thursday 25 April 2024  01:29:40 -0400 (0:00:00.186)       0:46:44.479 ******** 
=============================================================================== 
kustomize_deploy : Run Wait Conditions for examples/dt/uni01alpha ------------------------------------------------------------------------------------------------------------------------------------------------------------------- 768.64s
kustomize_deploy : Run Wait Conditions for examples/dt/uni01alpha/networker --------------------------------------------------------------------------------------------------------------------------------------------------------- 471.60s
kustomize_deploy : Run Wait Conditions for examples/dt/uni01alpha/control-plane ----------------------------------------------------------------------------------------------------------------------------------------------------- 450.02s
test_operator : Wait for the last job to be Completed - tempest --------------------------------------------------------------------------------------------------------------------------------------------------------------------- 396.85s
os_must_gather : Run openstack-must-gather command ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 185.09s
openshift_obs : Wait for observability-operator pod ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 60.61s
ci_setup : Install needed packages --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 54.03s
kustomize_deploy : Wait for cert-manager-operator pods ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 40.62s
kustomize_deploy : Wait for MetalLB speaker pods ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 36.04s
test_operator : Wait until the test-operator csv is present -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 33.56s
test_operator : Wait for the test-operator csv to Succeed ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 32.11s
openshift_setup : Wait for openshift-marketplace pods to be running ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 31.83s
kustomize_deploy : Wait for certmanager pods ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 30.93s
kustomize_deploy : Wait for NMstate webhook deployment ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 20.75s
kustomize_deploy : Wait for NMstate handler pods ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 20.66s
artifacts : Ensure we have proper rights on the gathered content --------------------------------------------------------------------------------------------------------------------------------------------------------------------- 20.00s
kustomize_deploy : Wait for cainjector pods ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 11.00s
test_operator : Ensure that the test-operator-logs-pod is Running -------------------------------------------------------------------------------------------------------------------------------------------------------------------- 10.53s
kustomize_deploy : Run Wait Conditions for examples/dt/uni01alpha/control-plane/nncp -------------------------------------------------------------------------------------------------------------------------------------------------- 6.57s
kustomize_deploy : Wait for webhook pods ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 5.52s
```

**Comment by fultonj** [comment]
Is this PR (197) being done in response to this change to the dataplane operator?

https://github.com/openstack-k8s-operators/dataplane-operator/pull/839

I think @abays added his most recent change to this patch (197) because he still getting stuck on the check introduced by PR839 above.

https://github.com/openstack-k8s-operators/architecture/compare/1c87769bd24b95cbc96003a5741fb20f3ec7e63c..eca199600cd164fde4550a398794e9eb9c91fb76

**Comment by fultonj** [comment]
> Is this PR (197) being done in response to this change to the dataplane operator?
> 
> [openstack-k8s-operators/dataplane-operator#839](https://github.com/openstack-k8s-operators/dataplane-operator/pull/839)
> 
> I think @abays added his most recent change to this patch (197) because he is still getting stuck on the check introduced by PR839 above.

That is fixed by https://github.com/openstack-k8s-operators/dataplane-operator/pull/859

**Review by abays — APPROVED** [review]
/lgtm

**Pull request merged**
